src/jit/flowgraph.cpp

```diff
@@ -5746,14 +5746,47 @@ void          Compiler::fgFindBasicBlocks()
 
     if (compIsForInlining())
     {
+        bool hasReturnBlocks = false;
+        bool hasMoreThanOneReturnBlock = false;
+
+        for (BasicBlock* block = fgFirstBB; block != nullptr; block = block->bbNext)
+        {
+            if (block->bbJumpKind == BBJ_RETURN)
+            {
+                if (hasReturnBlocks)
+                {
+                    hasMoreThanOneReturnBlock = true;
+                    break;
+                }
+
+                hasReturnBlocks = true;
+            }
+        }
+
+        if (!hasReturnBlocks)
+        {
+            //
+            // Mark the call node as "no return". The inliner might ignore CALLEE_DOES_NOT_RETURN and
+            // fail inline for a different reasons. In that case we still want to make the "no return"
+            // information available to the caller as it can impact caller's code quality.
+            //
+
+            impInlineInfo->iciCall->gtCallMoreFlags |= GTF_CALL_M_DOES_NOT_RETURN;
+
+            compInlineResult->NoteInt(InlineObservation::CALLEE_DOES_NOT_RETURN, static_cast<int>(fgBBcount));
```

Comment:
We currently don't fire off `NUMBER_OF_BASIC_BLOCKS` until after `fgFindBasicBlocks' returns.
Also LegacyPolicy does not remember the values of most observations, it tends to just act on things immediately.

So likely in the new policy you'll want to do two things: remember the value of the noreturn observation, and (in DEBUG) remember that the noreturn observation was made. Then when you get the basic block count observation, assert for the latter, then check the block count and the noreturn value and trigger the inlining failure.